### PR TITLE
Add Querier support for Graphs

### LINKED ISSUE #258: Support for querier component in frontend for variable time-based querying
The current querying in the frontend happens for recent 1-hour samples. The querier in br supports time-based querying that is capable of fetching samples within a given time-range. The UI for this is to be designed.

### LINKED ISSUE #383: The Journal Graph appears instead of the dialog box stating the metrics are loading
#### Describe
Although not easily visible if you decrease the data transfer speed using dev tools  it will be observed that the blank Journal graph apppears unlike the System metrics graph where a dialog box is displayed instead.

Screenshot from 2021-03-17 15-48-09

### PR DESCRIPTION
**Pull request by ruddi10**
Signed-off-by: ruddi <rudrakshaggarwalsachin@gmail.com>

fixes #258 
fixes #383 
### **Feature Added-**
Querier Support for  graphs.

Peek 2021-03-15 22-56

### CODE REVIEW

**Comment by ruddi10** [comment]
@Harkishen-Singh @muskankhedia have implemented the querier frontend of system metrics just wanted to ask one suggestion currently I have given the user to enter time only but we can give him option of time and date too so what do you think we should do?? Will implement accordingly

**Review by muskankhedia — CHANGES_REQUESTED** [review]
@ruddi10 This looks good. 

I would request you to make some changes:
1. We need to provide the user with the option to select both date and time. 
2. Along with this can you try implementing the [react-datetime](https://www.npmjs.com/package/react-datetime) picker which we are planning to implement in our revamped version of UI. You can continue with the next commit here itself, so if that didn't work well we can revert easily.
3. As we follow the unix date-time value to store the information and the implementation of conversion to unix values may differ with different languages. For the safer side, I would suggest you send the date-time in the raw format to the API and covert them to unix values at the backend. For this, you need to make some changes in the backend of the code, [here](https://github.com/bench-routes/bench-routes/blob/master/src/lib/api/api.go) is the implementation of API in golang.

**Comment by ruddi10** [comment]
@muskankhedia  unix conversion is working fine as of now so I think its consistent but still shall I do change in the backend code?

**Comment by ruddi10** [comment]
> @ruddi10 This looks good.
> 
> I would request you to make some changes:
> 
> 1. We need to provide the user with the option to select both date and time.
> 2. Along with this can you try implementing the [react-datetime](https://www.npmjs.com/package/react-datetime) picker which we are planning to implement in our revamped version of UI. You can continue with the next commit here itself, so if that didn't work well we can revert easily.
> 3. As we follow the unix date-time value to store the information and the implementation of conversion to unix values may differ with different languages. For the safer side, I would suggest you send the date-time in the raw format to the API and covert them to unix values at the backend. For this, you need to make some changes in the backend of the code, [here](https://www.npmjs.com/package/react-datetime) is the implementation of API in golang.

That api link is redirecting to react date time

**Comment by muskankhedia** [comment]
> > @ruddi10 This looks good.
> > I would request you to make some changes:
> > 
> > 1. We need to provide the user with the option to select both date and time.
> > 2. Along with this can you try implementing the [react-datetime](https://www.npmjs.com/package/react-datetime) picker which we are planning to implement in our revamped version of UI. You can continue with the next commit here itself, so if that didn't work well we can revert easily.
> > 3. As we follow the unix date-time value to store the information and the implementation of conversion to unix values may differ with different languages. For the safer side, I would suggest you send the date-time in the raw format to the API and covert them to unix values at the backend. For this, you need to make some changes in the backend of the code, [here](https://www.npmjs.com/package/react-datetime) is the implementation of API in golang.
> 
> That api link is redirecting to react date time

My bad, have updated it.

**Comment by ruddi10** [comment]
@muskankhedia is the failing of this test a problem as its working fine on my local env and one test is working fine

**Comment by ruddi10** [comment]
@muskankhedia  I have made the required changes.PTAL

**Comment by ruddi10** [comment]
Fix of #383 
The speed of fetching is a bit slow here due to my network settings so that the changes are visible

Peek 2021-03-18 18-42

**Comment by muskankhedia** [comment]
> Fix of #383
> The speed of fetching is a bit slow here due to my network settings so that the changes are visible
> 
> Peek 2021-03-18 18-42

@ruddi10 The implementation looks good. I will review the code this weekend.

**Comment by ruddi10** [comment]
@muskankhedia I guess due to some recent push on master this pr is having conflict so once you have reviewed the code let me know I will resolve that .

File: dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx [hunk 1]
```diff
@@ -141,7 +148,7 @@ const SystemMetrics: FC<SystemMetricsProps> = ({ showLoader }) => {
         }
       );
     // eslint-disable-next-line
```
Comment (Harkishen-Singh) [inline dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx:150]: This is valid/required anymore?

File: dashboard/v1.1/src/pages/Monitoring/Matrix.tsx [hunk 2]
```diff
@@ -46,32 +47,17 @@ const Element: FC<ElementProps> = ({
   const [trigger, setTrigger] = useState<number>(0);
   const [updating, setUpdating] = useState<boolean>(true);
   const [warning, showWarning] = useState<boolean>(false);
-  const endTimestamp = new Date().getTime() * 1000000 - TimeInstance.Hour;
+  // const endTimestamp = new Date().getTime() * 1000000 - TimeInstance.Hour;
```
Comment (Harkishen-Singh) [inline dashboard/v1.1/src/pages/Monitoring/Matrix.tsx:50]: debug comment

File: src/lib/api/api.go [hunk 3]
```diff
@@ -172,45 +172,41 @@ func (a *API) RoutesSummary(w http.ResponseWriter, _ *http.Request) {
...
-		err                          error
 	)
 	timeSeriesPath := r.URL.Query().Get("timeSeriesPath")
-
 	startTimestampStr := r.URL.Query().Get("startTimestamp")
 	if startTimestampStr == "" {
 		startTimestamp = int64(math.MaxInt64)
 	} else {
-		startTimestamp, err = strconv.ParseInt(startTimestampStr, 10, 64)
+		t, err := time.Parse(time.RFC3339, startTimestampStr)
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:181]: Can we make changes in UI to return in the format of [`http.TimeFormat`?](https://golang.org/pkg/net/http/#pkg-constants) That is some expected standard that we should do, but nothing hard rule here.

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
Thank you for the PR. Some questions, other than that, rest all looks good.

PS: It seems like a query is done on every time click, which should actually happen on clicking a execute button, but we aren't very aggressive in UI of v1.1

**Comment by ruddi10** [comment]
> Thank you for the PR. Some questions, other than that, rest all looks good.
> 
> PS: It seems like a query is done on every time click, which should actually happen on clicking a execute button, but we aren't very aggressive in UI of v1.1

@Harkishen-Singh we can implement that so shall I make request when the date-picker is closed?

File: dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx (hunk 1 above)
Comment (ruddi10) [reply dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx:150]: @Harkishen-Singh are you talking about showLoader props?

File: src/lib/api/api.go (hunk 3 above)
Comment (ruddi10) [reply src/lib/api/api.go:181]: Sure will do that

File: dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx (hunk 1 above)
Comment (Harkishen-Singh) [reply dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx:150]: About the `// eslint-disable-next-line`. Do we still need to ignore them?

**Comment by Harkishen-Singh** [comment]
> when the date-picker is closed?

Yes, or even better for I would suggest a submit button in the layout of start and end time, towards the right in a neat UI.

File: dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx (hunk 1 above)
Comment (ruddi10) [reply dashboard/v1.1/src/pages/Dashboard/SystemMetrics.tsx:150]: Yes

**Comment by ruddi10** [comment]
> > when the date-picker is closed?
> 
> Yes, or even better for I would suggest a submit button in the layout of start and end time, towards the right in a neat UI.

We can do that only problem is for that monitoring graph there we dont have much space.

**Comment by Harkishen-Singh** [comment]
If that's the case, then let's go with on close of datetime picker.

File: src/lib/api/api.go (hunk 3 above)
Comment (ruddi10) [reply src/lib/api/api.go:181]: @Harkishen-Singh unlike iso format http.TimeFormat dont have inbuilt pattern conversion in both moment js as well as golang so do you think we should convert it to that format by manually writting the pattern we want it to be converted to? or shall we continue with iso string only

**Comment by ruddi10** [comment]
@Harkishen-Singh  @muskankhedia kindly review

**Review by muskankhedia — APPROVED** [review]
LGTM!

**Comment by muskankhedia** [comment]
@ruddi10 Please rebase your branch with the master branch and squash all your commits to a single commit.

**Comment by ruddi10** [comment]
> @ruddi10 Please rebase your branch with the master branch and squash all your commits to a single commit.

@muskankhedia done

**Comment by muskankhedia** [comment]
@ruddi10 Please have a look at this error, this might be because the theme-mode feature has been modified now.
Screenshot 2021-03-27 at 2 21 47 PM

**Comment by muskankhedia** [comment]
@ruddi10 Please have a look at this error, this might be because the theme-mode approach has been modified recently.

Screenshot 2021-03-27 at 2 23 55 PM

**Comment by ruddi10** [comment]
> @ruddi10 Please have a look at this error, this might be because the theme-mode approach has been modified recently.
> 
> Screenshot 2021-03-27 at 2 23 55 PM

yes I have pushed again now missed that

**Comment by ruddi10** [comment]
> @ruddi10 Please have a look at this error, this might be because the theme-mode approach has been modified recently.
> Screenshot 2021-03-27 at 2 21 47 PM



> @ruddi10 Please have a look at this error, this might be because the theme-mode approach has been modified recently.
> Screenshot 2021-03-27 at 2 21 47 PM

@muskankhedia run yarn install for that

**Comment by ruddi10** [comment]
@muskankhedia I have fixed the errors kindly review

**Comment by muskankhedia** [comment]
Thank you for the PR!

**Pull request merged**